cmd/snap/complete.go

```diff
@@ -36,6 +37,95 @@ import (
 	"github.com/snapcore/snapd/snap"
 )
 
+// SnapAndApp holds a snap name and an application name
+type SnapAndApp struct {
+	Snap    string
+	App     string
+	has_dot bool
+}
+
+func (sa *SnapAndApp) FullName() string {
+	if sa.has_dot {
+		return fmt.Sprint(sa.Snap, ".", sa.App)
+	} else {
+		return sa.Snap
+	}
+}
+
+// UnmarshalFlag unmarshals the snap and application name. The following
+// combinations are allowed:
+// * <snap>.<app>
+// * <snap>
+// Every other combination results in an error.
+func (sa *SnapAndApp) UnmarshalFlag(value string) error {
+	parts := strings.Split(value, ".")
+	sa.Snap = ""
+	sa.App = ""
+	sa.has_dot = false
+	switch len(parts) {
+	case 1:
+		sa.Snap = parts[0]
+	case 2:
+		sa.Snap = parts[0]
+		sa.App = parts[1]
+		sa.has_dot = true // allows to know if it is "snap." or "snap.XXXX"
+	}
+	if sa.Snap == "" && sa.App != "" {
+		return fmt.Errorf(i18n.G("invalid value: %q (want snap.name or snap)"), value)
+	}
+	return nil
+}
+
+func (s SnapAndApp) Complete(match string) []flags.Completion {
+	var matchSnap SnapAndApp
+	matchSnap.UnmarshalFlag(match)
+	snaps, err := mkClient().List(nil, nil)
+	if err != nil {
+		return nil
+	}
+	if !matchSnap.has_dot {
+		// No dot in match, so complete with snap names
+		ret := make([]flags.Completion, 0, len(snaps))
+		for _, snap_t := range snaps {
```

Comment:
```suggestion
		for _, name := range snaps {
```

Reply:
Not sure... that's not the snap name, but a `client.Snap` structure/object. The name is a property inside it...

Reply:
(and can't use `snap` as the name because it would conflict with the import...)

Reply:
I like it. Done.